File: problem139.py
"""

dp[i] = True -----> dp[j] and s[j+1:i] exists in wordDict
      = False  ----> otherwise

"""
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        dp=[False]*(len(s)+1)
        wordSet = frozenset(wordDict)
        dp[0]=True
        for i in range(len(s)+1):
            for j in range(i):
                if dp[j] and s[j:i] in wordSet:
                    dp[i]=1
                    
        return dp[len(s)]
    
#Time: O(N)
#Space: O(N)

File: problem139.cs
public class Solution {
    string s;
    IList<string> wordDict;
    int[] dp;
    public bool WordBreak(string s, IList<string> wordDict) {
        this.s = s;
        this.wordDict = wordDict;
        this.dp = new int[s.Length];
        return FindSegment(0);
    }
    
    public bool FindSegment(int index)
    {
        if(index==s.Length)
        {
            return true;
        }
        
        if(dp[index]!=0)
        {
            return dp[index]==1?true:false;
        }
        
        StringBuilder sb = new StringBuilder();
        for(int i=index; i<s.Length; i++)
        {
            sb.Append(s[i]);
            if(!wordDict.Contains(sb.ToString()))
            {
                continue;
            }
            if(FindSegment(i+1))
            {
                dp[index]=1;
                return true;
            }
        }
        
        dp[index]=-1;
        return false;
    }
}